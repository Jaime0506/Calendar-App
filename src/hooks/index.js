export * from "./useUiStore"
export * from "./useForm"
export * from './useCalendarStore'
export * from './useAuthStore'