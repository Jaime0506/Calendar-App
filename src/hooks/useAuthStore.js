import { useDispatch, useSelector } from "react-redux"

// Verificar porque el thunks, normalmente se usa para esperar las respuestas de un tercero
// en este caso estamos consumiendo nuestra propia api

import { onCheckingToken, onLogin, onLogout, onRegister } from "../store/auth/thunks"
import { setErrorMessage } from "../store/auth/authSlice"

export const useAuthStore = () => {
    const { status, uid, user, errorMessage } = useSelector(store => store.auth)
    const dispatch = useDispatch()

    const handleOnLogin = (email, pasword) => {
        dispatch(onLogin(email,pasword))
    }

    const handleChecking = () => {

        const token = localStorage.getItem('token')

        if(token){
            dispatch(onCheckingToken())
        } else {
            dispatch(onLogout())
        }
    }

    const handleOnLogout = () => {
        dispatch(onLogout())
    }

    // No pasa por los thunks ya que es completamente sincronico
    const handleOnErrorForm = (error) => {
        dispatch(setErrorMessage(error))
    }

    const handleOnRegister = (name, email, password) => {
        dispatch(onRegister(name, email, password))
    }

    return {
        status,
        user,
        errorMessage,
        uid,
        handleOnLogin,
        handleChecking,
        handleOnLogout,
        handleOnErrorForm,
        handleOnRegister
    }
}