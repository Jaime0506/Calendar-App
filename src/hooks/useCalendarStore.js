import { useDispatch, useSelector } from "react-redux";
import { onDeleteEvent, onLoadingEvents, onNewEvent, onSetActiveEvent, onUpdateEvent } from "../store/calendar/thunks";

export const useCalendarStore = () => {
    const { events, activeEvent } = useSelector((store) => store.calendar);
    const dispatch = useDispatch()

    const handleOnSetActiveEvent = (calendarEvent) => {
        dispatch(onSetActiveEvent(calendarEvent))
    };

    const handleOnSavingEvent = (calendarEvent, color) => {
        if (calendarEvent.id) {
            // SI LA RESPUESTA DEL BACKEND TIENE ID, SIGNIFICA QUE ESTAMOS
            // ACTUALIZANDO UN EVENTO
            dispatch(onUpdateEvent({ ...calendarEvent, bgColor: color }))
        } else {
            // DE LO CONTRARIO SIGNIFICA QUE ESTAMOS CREANDO UN EVENTO Y DEBEMOS
            // ASIGNARLE UN _ID A TRAVES DEL BACKEND
            dispatch(onNewEvent({...calendarEvent, color }))
        }
    }

    const handleOnDeleteEvent = () => {
        dispatch(onDeleteEvent())
    }

    const handleOnLoadingEvents = () => {
        dispatch(onLoadingEvents())
    }

    return {
        events,
        activeEvent,
        hasOnEventSelected: !!activeEvent,

        handleOnSetActiveEvent,
        handleOnSavingEvent,
        handleOnDeleteEvent,
        handleOnLoadingEvents
    };
};
