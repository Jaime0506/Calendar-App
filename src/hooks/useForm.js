import { differenceInSeconds } from "date-fns";
import { useState } from "react";

export const useForm = (initialState = {}) => {
    const [formValues, setFormValues] = useState(initialState);
    const [error, setError] = useState(null)

    const onInputChange = ({ target }) => {
        setFormValues({
            ...formValues,
            [target.name]: target.value,
        });
    };

    const onPickerDateChange = (type, date) => {
        setFormValues({
            ...formValues,
            [type]: date,
        });
    };

    const validateForm = () => {
        setError(null)

        const difference = differenceInSeconds(
            formValues.end,
            formValues.start
        );


        if (formValues.start <= 0) {
            setError({ start: true })
            return {
                role: "start",
                message: "Los campos de fecha no pueden estar vacios",
            };
        }

        if (formValues.end <= 0) {
            setError({ end: true })
            return {
                role: "end",
                message: "Los campos de fecha no pueden estar vacios",
            };
        }

        if (formValues.title.length <= 0) {
            setError({ title: true })
            return {
                role: "title",
                message: "El titulo es obligatorio",
            };
        }

        if (difference <= 0 || isNaN(difference)) {
            setError({ end: true })
            return {
                role: "end",
                message: "la fecha de finalizacion debe ser mayor a la de inicio",
            };
        }
    };
    
    return {
        formValues,
        onInputChange,
        onPickerDateChange,
        validateForm,
        error
    };
};
