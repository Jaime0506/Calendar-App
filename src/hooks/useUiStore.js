import { useDispatch, useSelector } from "react-redux"
import { onCloseModal, onOpenModal } from "../store/ui"

export const useUiStore = () => {
    const { isVisibleToastify, messageToastify, isVisibleModal } = useSelector(state => state.ui)
    const dispatch = useDispatch()

    const openDateModal = () => {
        dispatch(onOpenModal())
    }

    const closeDateModal = () => {
        dispatch(onCloseModal())
    }

    return {
        // properties
        isVisibleToastify,
        messageToastify,
        isVisibleModal,

        openDateModal,
        closeDateModal,
    }
}