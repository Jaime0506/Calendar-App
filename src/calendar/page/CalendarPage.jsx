import { useState } from "react";
import { Calendar } from "react-big-calendar";

import { useUiStore, useCalendarStore } from "../../hooks";
import { localizer, getMessages } from "../../helpers";

import { Navbar, CalendarEvent, CalendarModal, FabAddNew, FabDelete } from "../";

import "react-big-calendar/lib/css/react-big-calendar.css";
import { useEffect } from "react";

export const CalendarPage = () => {
    // Guardamos en el locale storage, la ultima vista visitada
    const { openDateModal, isVisibleModal } = useUiStore()
    const { events, handleOnSetActiveEvent, handleOnLoadingEvents } = useCalendarStore()

    const [lastView, setLastView] = useState(
        localStorage.getItem("lastView") || "week"
    );

    const onDobleClick = () => {
        openDateModal()
    };

    const onSelected = (event) => {
        handleOnSetActiveEvent(event)
    };

    const onViewChanged = (event) => {
        localStorage.setItem("lastView", event);
        setLastView(event);
    };    

    useEffect(() => {
        handleOnLoadingEvents()
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    // eslint-disable-next-line no-unused-vars
    const eventStyleGetter = (event, start, end, isSelected) => {
        const style = {
            backgroundColor: event.bgColor,
            color: "white",
        };

        return {
            style,
        };
    };

    return (
        <>
            <Navbar />
            <Calendar
                culture="es"
                events={events}
                defaultView={lastView}
                localizer={localizer}
                startAccessor="start"
                endAccessor="end"
                style={{ height: 'calc( 100vh - 80px )' }}
                messages={getMessages()}
                eventPropGetter= {eventStyleGetter }
                components={{
                    event: CalendarEvent,
                }}
                onDoubleClickEvent={onDobleClick}
                onSelectEvent={onSelected}
                onView={onViewChanged}
            />
            { isVisibleModal && <CalendarModal />}
            <FabAddNew />
            <FabDelete />
        </>
    );
};
