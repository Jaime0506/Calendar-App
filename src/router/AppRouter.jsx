import { Routes, Route, Navigate } from "react-router-dom";
import { LoginPage } from "../auth";
import { CalendarPage } from "../calendar";
import { useEffect } from "react";
import { useAuthStore } from "../hooks";
import { Checking } from "../components";

export const AppRouter = () => {

    const { status, handleChecking } = useAuthStore()

    useEffect(() => {
        handleChecking()
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    if (status == "checking") return <Checking />

    return (
        <Routes>
            {
                (status === "not-authenticated"
                    ? (
                        <>
                            <Route path="/auth/*" element={<LoginPage />} />
                            <Route path="/*" element={<Navigate to='/auth/login' />} />
                        </>
                    ) : (
                        <>
                            <Route path="/" element={<CalendarPage />} />)
                            <Route path="/*" element={<Navigate to='/' />} />
                        </>
                    )
                )
            }
            <Route path="/*" element={<Navigate to="/auth/login" />} />
        </Routes>
    );
};
