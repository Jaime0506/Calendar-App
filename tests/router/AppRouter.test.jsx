import { render, screen } from "@testing-library/react"
import { useAuthStore } from "../../src/hooks/useAuthStore"
import { AppRouter } from "../../src/router/AppRouter"
import { MemoryRouter } from "react-router-dom"
import { CalendarPage } from "../../src/calendar/page/CalendarPage"

jest.mock("../../src/hooks/useAuthStore")

// Simplemente necesitamos verificar que se renderize el componente de CalendarPage, cuando
// se encuentra authenticado, pero si renderizamos x defecto nos da muchos errores
// por los diferentes custom hooks que usamos en este componente, por esto
// Simplemente podemos hacer un mock del componente y que regrese un simple <h1>Calendar Page</h1>
jest.mock("../../src/calendar/page/CalendarPage", () => ({
    CalendarPage: () => <h1>Calendar Page</h1>
}))

describe('Testing in <AppRouter />', () => {

    const mockHandleChecking = jest.fn()

    beforeEach(() => jest.clearAllMocks())

    test('the Loading should be displayed and the handleCheckin function should be called when the component is mounted', () => {

        useAuthStore.mockReturnValue({
            status: "checking",
            handleChecking: mockHandleChecking
        })

        render(<AppRouter />)

        expect(screen.getByText("Loading...")).toBeTruthy()
        expect(mockHandleChecking).toHaveBeenCalled()
    })

    test('should show <LoginPage /> when status is not-authenticated', () => {

        useAuthStore.mockReturnValue({
            status: "not-authenticated",
            handleChecking: mockHandleChecking
        })

        const { container } = render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>
        )
        
        expect(screen.getByText("Ingreso")).toBeTruthy()
        expect(container).toMatchSnapshot()
    })

    test('should show <CalendarPage /> when status is authenticated', () => {  
        useAuthStore.mockReturnValue({
            status: "authenticated",
            handleChecking: mockHandleChecking
        })

        const { container } = render(
            <MemoryRouter>
                <AppRouter />
            </MemoryRouter>
        )

        expect(screen.getByText("Calendar Page")).toBeTruthy()
        expect(container).toMatchSnapshot()
    })
})