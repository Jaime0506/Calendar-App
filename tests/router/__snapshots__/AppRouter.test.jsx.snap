// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <AppRouter /> should show <CalendarPage /> when status is authenticated 1`] = `
<div>
  <h1>
    Calendar Page
  </h1>
</div>
`;

exports[`Testing in <AppRouter /> should show <LoginPage /> when status is not-authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="email"
              placeholder="Correo"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="password"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2 mt-3"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control null"
              name="name"
              placeholder="Nombre"
              type="text"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control null"
              name="email"
              placeholder="Correo"
              type="email"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control null"
              name="password"
              placeholder="Contraseña"
              type="password"
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="repeatPassword"
              placeholder="Repita la contraseña"
              type="password"
            />
          </div>
          <div
            class="d-grid gap-2 mt-3"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Crear cuenta"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
